### PR TITLE
Admins can import shows via the NYT Events API

### PR DESCRIPTION
**Pull request by reshleman**
**Implement NYT Events API wrapper:**
- Define `NytEventListings` module
- Define `Event`, `Query`, and `Result` classes within module
- Update Gemfile with `dotenv-rails` and `httparty`
- Add `.env.sample` with sample NYT Events API key
- Query the API on `events#index`

**Admins can import shows via the NYT Events API**
- Add `nyt_event_id` and `nyt_updated_at` columns to `events`
- Add validation on `nyt_event_id`
- Create rake task for importing events from the API
- Create `EventsImporter` class to do the hard work
- Add helper method to `Event`

https://trello.com/c/nzVFiyco

### CODE REVIEW

File: app/models/nyt_event_listings/event.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+      end
+    end
+
+    def method_missing(name, *args)
+      if respond_to?(name)
+        @fields[name.to_s]
+      else
+        super
+      end
+    end
```
Comment (hrs) [inline app/models/nyt_event_listings/event.rb]: Nice `method_missing`, but you might look into [OpenStructs](http://www.ruby-doc.org/stdlib-2.1.2/libdoc/ostruct/rdoc/OpenStruct.html) instead. They do pretty much exactly what you're doing here.

Comment (reshleman) [inline app/models/nyt_event_listings/event.rb]: Ah, good call. This class is basically just a wrapper for a hash anyway.

File: app/models/nyt_event_listings/result.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module NytEventListings
+  class Result
+    attr_reader :status, :copyright, :num_results, :results
+    alias_method :events, :results
+
+    def initialize(query_result)
+      @status = query_result["status"]
+      @copyright = query_result["copyright"]
+      @num_results = query_result["num_results"]
```
Comment (hrs) [inline app/models/nyt_event_listings/result.rb]: How much do you care about this metadata? If the answer is "not much", I'd drop this class and have `Query#execute` return an array of Events.

File: app/views/events/index.html.erb [hunk 3]
```diff
@@ -9,3 +9,15 @@
...
   <%= render @closed_events %>
 </div>
+
+<div>
+  <h2>NYT Events</h2>
+
+  <%= @nyt_events_query.status %><br />
+  <%= @nyt_events_query.num_results %><br />
+  <%= @nyt_events_query.events.inspect %><br />
+  <% @nyt_events_query.events.each do |event| %>
```
Comment (halogenandtoast) [inline app/views/events/index.html.erb]: I'd throw a newline above this just to increase readability.

File: app/models/nyt_events/query.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        NytEvents::Event.new(event)
+      end
+    end
+
+    def status
+      @query_result.response.code.to_i
+    end
+
+    def num_results
+      @query_result["num_results"]
```
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: I see you have an attr_reader below, any reason not to use it.

File: app/models/nyt_events/query.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
+    def events
+      @query_result.fetch("results", []).map do |event|
+        NytEvents::Event.new(event)
+      end
+    end
+
+    def status
+      @query_result.response.code.to_i
```
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: see below.

File: app/models/nyt_events/query.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+module NytEvents
+  class Query
+    def initialize(options)
+      @options = options
+      @query_result = perform_query
+    end
+
+    def events
+      @query_result.fetch("results", []).map do |event|
```
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: see even further below.

File: app/models/nyt_events/query.rb (hunk 4 above)
Comment (reshleman) [inline app/models/nyt_events/query.rb]: Good catch, thanks.

File: app/models/nyt_events/query.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    private
+
+    attr_reader :options, :query_result
+
+    def perform_query
+      HTTParty::get(URL, query_params)
+    end
+
+    def query_params
```
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: seems useless, why not just set this in the initialize.

File: app/models/nyt_events/query.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      { query: @options.merge("api-key" => API_KEY) }
```
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: Could also pull out the api-key hash into a DEFAULT_OPTIONS constant.

Comment (reshleman) [inline app/models/nyt_events/query.rb]: @halogenandtoast Would you define that in the module w/ the other constants or in the class?

File: app/models/nyt_events/query.rb (hunk 7 above)
Comment (reshleman) [inline app/models/nyt_events/query.rb]: :+1:

File: app/models/nyt_events/query.rb (hunk 8 above)
Comment (halogenandtoast) [inline app/models/nyt_events/query.rb]: I'd define it in this class because that's who care about it.

File: app/helpers/date_helper.rb [hunk 9]
```diff
@@ -1,6 +1,8 @@
 module DateHelper
   def closing_date_in_words(date)
-    if still_open?(date)
+    if date.nil?
```
Comment (hrs) [inline app/helpers/date_helper.rb]: I might _even_ break this into a nicely-named helper method. Maybe `no_closing_date?(date)`?

File: app/models/event.rb [hunk 10]
```diff
@@ -3,12 +3,17 @@ class Event < ActiveRecord::Base
...
   def self.current
-    where("closing_date >= ?", Date.today)
+    where("closing_date IS NULL OR closing_date >= ?", Date.today)
   end
 
   def self.closed
     where("closing_date < ?", Date.today)
   end
+
+  def older_than_nyt?(date)
```
Comment (hrs) [inline app/models/event.rb]: I'm not sure this reads well. `event.older_than_nyt?(date)`

Research reveals that to be older than the NYT an event must have taken place before Sept. 18, 1851.

File: app/models/events_importer.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  def import_event(source)
+    event = Event.find_or_initialize_by(nyt_event_id: source.event_id)
+
+    if event.new_record? || event.older_than_nyt?(source.last_modified)
+      update_or_create_from(event, source)
+    end
+  end
+
+  def update_or_create_from(event, source)
+    event.name = strip_single_quotes(source.event_name)
```
Comment (hrs) [inline app/models/events_importer.rb]: Maybe I'm misreading this, but is this the right thing to do? What about apostrophes? Wouldn't you want those in an event name?

File: app/models/events_importer.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    event.nyt_updated_at = source.last_modified
+    event.save!
+  end
+
+  def strip_single_quotes(string)
+    string.strip.gsub(/\A‘|’\Z/, "")
+  end
+
+  def strip_p_tags(string)
+    string.strip.gsub(/\A<p>|<\/p>\Z/, "")
```
Comment (hrs) [inline app/models/events_importer.rb:46]: Using [String#tr](http://ruby-doc.org/core-2.1.2/String.html#method-i-tr) might read better.

File: app/models/events_importer.rb (hunk 11 above)
Comment (reshleman) [inline app/models/events_importer.rb]: @hrs It removes the surrounding single quotes. I named it similarly to `String#strip` which removes surrounding (but not embedded) whitespace.

I'm open to renaming if you think that's confusing. I previously had something like `remove_surrounding_single_quotes` or `strip_surrounding_single_quotes`.

Comment (hrs) [inline app/models/events_importer.rb]: `strip_surrounding_single_quotes` might be more clear. Should you be stripping double quotes, too? `strip_surrounding_quotes` would be even nicer.

Comment (reshleman) [inline app/models/events_importer.rb]: Eh, maybe. Right now I'm just handling the specific cases from the API data, which has single left and right smart quotes around the event titles. >_<

Comment (reshleman) [inline app/models/events_importer.rb]: I'm debating if it's worth using a gem. There are Unicode characters in other fields that are giving me problems, too.

Comment (hrs) [inline app/models/events_importer.rb]: Hmm. Maybe? If it's a big impediment, I'd search for a gem. But this is also something you can easily change later, I think.

Comment (reshleman) [inline app/models/events_importer.rb]: My thought as well. It's not blocking me from moving on, and I'd like start on another feature. I think I'll add a separate Trello card for it and come back to it later.

Comment (hrs) [inline app/models/events_importer.rb]: :thumbsup:

**Comment by hrs** [comment]
:dango:

**Pull request closed without merging**